File: tool/test/mandelbrot.c
int main() {
    int X_SIZE = 10;
    int Y_SIZE = 10;
    float image[Y_SIZE][X_SIZE];

    float CENTER_X = -0.5;
    float CENTER_Y = 0.0;
    float SCALE_X = 2.0;
    float SCALE_Y = 2.0;
    float y = 0, x = 0;

    y = CENTER_Y + SCALE_Y / 2.0;
    float dy = SCALE_Y / Y_SIZE;
    float max = 0;
    for (int i = 0; i < Y_SIZE; i++) {
        y -= dy;
        x = CENTER_X - SCALE_X / 2.0;
        float dx = SCALE_X / X_SIZE;
        for (int j = 0; j < X_SIZE; j++) {
            x += dx;
            image[i][j] = 0;
            float z1 = 0, z2 = 0, z1_ = 0, z2_ = 0;
            for (int n = 0; n < 20; n++) {
                z1_ = z1 * z1 - z2 * z2 + x;
                z2_ = 2 * z1 * z2 + y;
                if (z1_ * z1_ + z2_ * z2_ > 4.0) {
                    image[i][j] = n;
                    if (image[i][j] > max) {
                        max = image[i][j];
                    }
                    break;
                }
                z1 = z1_;
                z2 = z2_;
            }
        }
    }

    int image_[Y_SIZE][X_SIZE];
    for (int i = 0; i < Y_SIZE; i++) {
        for (int j = 0; j < X_SIZE; j++) {
            image_[i][j] = (int)((image[i][j] / max) * 255);
        }
    }

    return 0;
}

File: tool/test/mandelbrot.py
import numpy as np
import math
import matplotlib.pyplot as plt

center = (-0.5, 0)
scale = (2.0, 2.0)


def mandelbrot(x, y):
    z1, z2 = 0, 0
    for n in range(1, 20):
        z1, z2 = z1 * z1 - z2 * z2 + x, 2 * z1 * z2 + y
        if (z1 * z1 + z2 * z2) > 4.0:
            return n

    return 0


def calc_image(x_size, y_size):
    image = np.zeros((y_size, x_size), dtype=np.float32)

    y = center[1] + scale[1] / 2.0
    dy = scale[1] / y_size
    for i in range(y_size):
        y -= dy
        x = center[0] - scale[0] / 2.0
        dx = scale[0] / x_size
        for j in range(x_size):
            x += dx
            image[i, j] = mandelbrot(x, y)

    return image


def main():
    image = calc_image(10, 10)
    plt.imshow(image, cmap="gray")
    plt.show()


if __name__ == "__main__":
    main()
